README.md
# bip66

Strict DER signature encoding/decoding.

See [bip66](https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki).

- This module **works only with [two's complement](https://en.wikipedia.org/wiki/Two's_complement) numbers**.
- BIP66 doesn't check that `r` or `s` are fully valid.
    - `check`/`decode` doesn't check that `r` or `s` great than 33 bytes or that this number represent valid point on elliptic curve.
    - `encode` doesn't check that `r`/`s` represent valid point on elliptic curve.

## Example

``` dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_bip66/dart_bip66.dart';

void main() {
  Uint8List r = Uint8List.fromList(hex.decode(
      '1ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777'));
  Uint8List s = Uint8List.fromList(hex.decode(
      '29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1'));
  // => 304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1
  print(der.encode(DER(r: r, s: s)));
  // => [48, 67, 2, 32, 30, 161, 253, 255, 129, 179, 162, 113, 101, 157, 244, 170, 209, 155, 196, 239, 131, 222, 243, 137, 19, 26, 54, 53, 143, 230, 75, 36, 86, 50, 231, 119, 2, 31, 41, 225, 100, 101, 139, 233, 206, 129, 9, 33, 191, 129, 214, 184, 102, 148, 120, 90, 121, 234, 30, 82, 219, 250, 81, 5, 20, 141, 31, 11, 193]

  Uint8List signature = Uint8List.fromList(hex.decode(
      '304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1'));

  DER decodedSignature = der.decode(signature);
  print(decodedSignature);
  // => 304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1
  print(der.encode(decodedSignature));
  // => [48, 67, 2, 32, 30, 161, 253, 255, 129, 179, 162, 113, 101, 157, 244, 170, 209, 155, 196, 239, 131, 222, 243, 137, 19, 26, 54, 53, 143, 230, 75, 36, 86, 50, 231, 119, 2, 31, 41, 225, 100, 101, 139, 233, 206, 129, 9, 33, 191, 129, 214, 184, 102, 148, 120, 90, 121, 234, 30, 82, 219, 250, 81, 5, 20, 141, 31, 11, 193]
}
```

## LICENSE [MIT](LICENSE)

CHANGELOG.md
## 2.0.1

- update readme for example

## 2.0.0

- update this package for Codec and Converter use

## 1.0.0

- Initial version.

test/fixtures.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';

enum FixtureEnum { valid, invalidEncode, invalidDecode }

class Fixture {
  String? der;
  Uint8List? get derBuffer {
    return Uint8List.fromList(hex.decode(der!));
  }

  String? r;
  String? s;
  String? exception;

  Fixture({this.der, this.r, this.s, this.exception});
}

Map<FixtureEnum, List<Fixture>> fixtures = {
  FixtureEnum.valid: [
    Fixture(
        der:
            "3044022029db2d5f4e1dcc04e19266cce3cb135865784c62ab653b307f0e0bb744f5c2aa022000a97f5826912cac8b44d9f577a26f169a2f8db781f2ddb7de2bc886e93b6844",
        r: "29db2d5f4e1dcc04e19266cce3cb135865784c62ab653b307f0e0bb744f5c2aa",
        s: "00a97f5826912cac8b44d9f577a26f169a2f8db781f2ddb7de2bc886e93b6844"),
    Fixture(
        der:
            "304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1",
        r: "1ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777",
        s: "29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1"),
    Fixture(
        der:
            "304402201b19519b38ca1e6813cd25649ad36be8bc6a6f2ad9758089c429acd9ce0b572f02203bf32193c8a3a3de1f847cd6e6eebf43c96df1ffa4d7fe920f8f71708920c65f",
        r: "1b19519b38ca1e6813cd25649ad36be8bc6a6f2ad9758089c429acd9ce0b572f",
        s: "3bf32193c8a3a3de1f847cd6e6eebf43c96df1ffa4d7fe920f8f71708920c65f"),
    Fixture(
        der:
            "3044022000c8da1836747d05a6a3d2c395825edce827147d15909e66939a5037d5916e6f022017823c2da62f539d7f8e1e186eaea7a401ab3c077dcfc44aeaf3e13fac99bdbc",
        r: "00c8da1836747d05a6a3d2c395825edce827147d15909e66939a5037d5916e6f",
        s: "17823c2da62f539d7f8e1e186eaea7a401ab3c077dcfc44aeaf3e13fac99bdbc"),
    Fixture(
        der:
            "3042021e2ff2609c8dc0392d3731a2c6312841e09c76f10b83e2b52604dc84886dd502200090ac80e787c063618192bc04758e6666d0179c377fb2f3d6105d58000f33ac",
        r: "2ff2609c8dc0392d3731a2c6312841e09c76f10b83e2b52604dc84886dd5",
        s: "0090ac80e787c063618192bc04758e6666d0179c377fb2f3d6105d58000f33ac"),
    Fixture(
        der:
            "3041021d00feb1a12c27e5fe261acc64c0923add082573883e0800d8e4080fa9bb02202e7aeb97f4046ea3be60d2896a19c8dc269ab5eb2de968912cd52a076a0a42e9",
        r: "00feb1a12c27e5fe261acc64c0923add082573883e0800d8e4080fa9bb",
        s: "2e7aeb97f4046ea3be60d2896a19c8dc269ab5eb2de968912cd52a076a0a42e9"),
    Fixture(
        der:
            "3042021d00feb1a12c27e5fe261acc64c0923add082573883e0800d8e4080fa9bb0221008aeb97f4046ea3be60d2896a19c8dc269ab5eb2de968912cd52a076a0a42e925",
        r: "00feb1a12c27e5fe261acc64c0923add082573883e0800d8e4080fa9bb",
        s: "008aeb97f4046ea3be60d2896a19c8dc269ab5eb2de968912cd52a076a0a42e925"),
    Fixture(
        der:
            "303e021d00c1d545da2e4edfbc65e9267d3c0a6fdda41793d0fd945f15acbcf0dd021d009acffda3ca5e7c349c35ba606f0a8f1ec7815b653b51695ca9ee69a6",
        r: "00c1d545da2e4edfbc65e9267d3c0a6fdda41793d0fd945f15acbcf0dd",
        s: "009acffda3ca5e7c349c35ba606f0a8f1ec7815b653b51695ca9ee69a6"),
    Fixture(der: "3006020100020100", r: "00", s: "00")
  ],
  FixtureEnum.invalidEncode: [
    Fixture(
        exception: "Exception: R length is zero",
        r: "",
        s: "0080000000000000000000000000000000000000000000000000000000000000"),
    Fixture(
        exception: "Exception: S length is zero",
        r: "0080000000000000000000000000000000000000000000000000000000000000",
        s: ""),
    Fixture(exception: "Exception: R value is negative", r: "80", s: "7f"),
    Fixture(exception: "Exception: S value is negative", r: "7f", s: "80"),
    Fixture(
        exception: "Exception: R value excessively padded",
        r: "0010000000000000000000000000000000000000000000000000000000000000",
        s: "0080000000000000000000000000000000000000000000000000000000000000"),
    Fixture(
        exception: "Exception: S value excessively padded",
        r: "0080000000000000000000000000000000000000000000000000000000000000",
        s: "0010000000000000000000000000000000000000000000000000000000000000"),
    Fixture(
        exception: "Exception: R length is too long",
        r: "00800000000000000000000000000000000000000000000000000000000000000000",
        s: "008000000000000000000000000000000000000000000000000000000000000000"),
    Fixture(
        exception: "Exception: S length is too long",
        r: "008000000000000000000000000000000000000000000000000000000000000000",
        s: "00800000000000000000000000000000000000000000000000000000000000000000")
  ],
  FixtureEnum.invalidDecode: [
    Fixture(
        exception: "Exception: DER sequence length is too short",
        der: "ffffffffffffff"),
    Fixture(
        exception: "Exception: DER sequence length is too long",
        der:
            "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"),
    Fixture(
        exception: "Exception: Expected DER sequence",
        der: "00ffff0400ffffff020400ffffff"),
    Fixture(
        exception: "Exception: DER sequence length is invalid",
        der: "30ff020400ffffff020400ffffff"),
    Fixture(
        exception: "Exception: DER sequence length is invalid",
        der: "300c030400ffffff030400ffffff0000"),
    Fixture(
        exception: "Exception: Expected DER integer",
        der: "300cff0400ffffff020400ffffff"),
    Fixture(
        exception: "Exception: Expected DER integer (2)",
        der: "300c020200ffffff020400ffffff"),
    Fixture(
        exception: "Exception: R length is zero", der: "30080200020400ffffff"),
    Fixture(
        exception: "Exception: S length is zero", der: "3008020400ffffff0200"),
    Fixture(
        exception: "Exception: R length is too long",
        der: "300c02dd00ffffff020400ffffff"),
    Fixture(
        exception: "Exception: S length is invalid",
        der: "300c020400ffffff02dd00ffffff"),
    Fixture(
        exception: "Exception: R value is negative",
        der: "300c020480000000020400ffffff"),
    Fixture(
        exception: "Exception: S value is negative",
        der: "300c020400ffffff020480000000"),
    Fixture(
        exception: "Exception: R value excessively padded",
        der: "300c02040000ffff020400ffffff"),
    Fixture(
        exception: "Exception: S value excessively padded",
        der: "300c020400ffffff02040000ffff")
  ]
};


test/dart_bip66_test.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_bip66/dart_bip66.dart';
import 'package:test/test.dart';

import 'fixtures.dart';

void main() {
  if (fixtures.containsKey(FixtureEnum.valid)) {
    for (Fixture fixture in fixtures[FixtureEnum.valid]!) {
      Uint8List derBuffer = fixture.derBuffer!;
      test('check: returns true for ${fixture.der}', () {
        expect(der.check(derBuffer), isTrue);
      });
      test('decode: ${fixture.der}', () {
        var signature = der.decode(derBuffer);
        expect(hex.encode(signature.r), equals(fixture.r));
        expect(hex.encode(signature.s), equals(fixture.s));
      });
      test('encode: ${fixture.r}, ${fixture.s}', () {
        Uint8List r = Uint8List.fromList(hex.decode(fixture.r!));
        Uint8List s = Uint8List.fromList(hex.decode(fixture.s!));
        DER der = DER(r: r, s: s);
        expect(der.toString(), equals(fixture.der));
      });
    }
  }
  if (fixtures.containsKey(FixtureEnum.invalidDecode)) {
    for (Fixture fixture in fixtures[FixtureEnum.invalidDecode]!) {
      Uint8List derBuffer = fixture.derBuffer!;
      test('check: returns false for ${fixture.der} (${fixture.exception})',
          () {
        expect(der.check(derBuffer), isFalse);
      });
      test('throws "Exception: ${fixture.exception}" for ${fixture.der}', () {
        DER? signature;
        try {
          signature = der.decode(derBuffer);
        } catch (e) {
          expect((e as Exception).toString(), equals(fixture.exception));
        }
        expect(signature, isNull);
      });
    }
  }
  if (fixtures.containsKey(FixtureEnum.invalidEncode)) {
    for (Fixture fixture in fixtures[FixtureEnum.invalidEncode]!) {
      test(
          'throws "Exception: ${fixture.exception}" for ${fixture.r}, ${fixture.s}',
          () {
        Uint8List r = Uint8List.fromList(hex.decode(fixture.r!));
        Uint8List s = Uint8List.fromList(hex.decode(fixture.s!));
        Uint8List? buff;
        try {
          buff = der.encode(DER(r: r, s: s));
        } catch (e) {
          expect((e as Exception).toString(), equals(fixture.exception));
        }
        expect(buff, isNull);
      });
    }
  }
}


lib/src/dart_bip66_base.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dart_bip66/src/converters/api.dart';
import 'package:dart_bip66/src/models/der_model.dart';

export 'package:dart_bip66/src/converters/api.dart' show derEncoder, derDecoder;
export 'package:dart_bip66/src/models/der_model.dart';

const der = DERCodec();

/// Strict DER signature encoding/decoding codec.
/// Reference https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
/// Format: 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S]
/// NOTE: SIGHASH byte ignored AND restricted, truncate before use
class DERCodec extends Codec<DER, Uint8List> {
  const DERCodec();

  /// Signature Check
  bool check(Uint8List buffer) {
    if (buffer.length < 8) return false;
    if (buffer.length > 72) return false;
    if (buffer[0] != 0x30) return false;
    if (buffer[1] != buffer.length - 2) return false;
    if (buffer[2] != 0x02) return false;

    var lenR = buffer[3];
    if (lenR == 0) return false;
    if (5 + lenR >= buffer.length) return false;
    if (buffer[4 + lenR] != 0x02) return false;

    var lenS = buffer[5 + lenR];
    if (lenS == 0) return false;
    if ((6 + lenR + lenS) != buffer.length) return false;

    if ((buffer[4] & 0x80) != 0) return false;
    if (lenR > 1 && (buffer[4] == 0x00) && !((buffer[5] & 0x80) != 0)) {
      return false;
    }

    if ((buffer[lenR + 6] & 0x80) != 0) return false;
    if (lenS > 1 &&
        (buffer[lenR + 6] == 0x00) &&
        !((buffer[lenR + 7] & 0x80) != 0)) return false;
    return true;
  }

  @override
  DerDecoder get decoder => derDecoder;

  @override
  DerEncoder get encoder => derEncoder;
}


lib/src/converters/decoder.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dart_bip66/src/models/der_model.dart';

/// The canonical instance of [DerDecoder].
const derDecoder = DerDecoder();

/// DER signature decoder
class DerDecoder extends Converter<Uint8List, DER> {
  const DerDecoder();

  @override
  DER convert(Uint8List input) {
    if (input.length < 8) throw Exception('DER sequence length is too short');
    if (input.length > 72) throw Exception('DER sequence length is too long');
    if (input[0] != 0x30) throw Exception('Expected DER sequence');
    if (input[1] != input.length - 2) {
      throw Exception('DER sequence length is invalid');
    }
    if (input[2] != 0x02) throw Exception('Expected DER integer');

    var lenR = input[3];
    if (lenR == 0) throw Exception('R length is zero');
    if (5 + lenR >= input.length) throw Exception('R length is too long');
    if (input[4 + lenR] != 0x02) throw Exception('Expected DER integer (2)');

    var lenS = input[5 + lenR];
    if (lenS == 0) throw Exception('S length is zero');
    if ((6 + lenR + lenS) != input.length) {
      throw Exception('S length is invalid');
    }

    if ((input[4] & 0x80) != 0) throw Exception('R value is negative');
    if (lenR > 1 && (input[4] == 0x00) && (input[5] & 0x80) == 0) {
      throw Exception('R value excessively padded');
    }

    if ((input[lenR + 6] & 0x80) != 0) throw Exception('S value is negative');
    if (lenS > 1 &&
        (input[lenR + 6] == 0x00) &&
        (input[lenR + 7] & 0x80) == 0) {
      throw Exception('S value excessively padded');
    }

    // non-BIP66 - extract R, S values
    return DER(r: input.sublist(4, 4 + lenR), s: input.sublist(6 + lenR));
  }
}


lib/src/converters/encoder.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dart_bip66/src/models/der_model.dart';

/// The canonical instance of [DerEncoder].
const derEncoder = DerEncoder();

/// DER signature encoder
class DerEncoder extends Converter<DER, Uint8List> {
  const DerEncoder();

  @override
  Uint8List convert(DER input) {
    int lenR = input.r.length;
    int lenS = input.s.length;
    if (lenR == 0) throw Exception('R length is zero');
    if (lenS == 0) throw Exception('S length is zero');
    if (lenR > 33) throw Exception('R length is too long');
    if (lenS > 33) throw Exception('S length is too long');
    if ((input.r[0] & 0x80) != 0) throw Exception('R value is negative');
    if ((input.s[0] & 0x80) != 0) throw Exception('S value is negative');
    if (lenR > 1 && (input.r[0] == 0x00) && !((input.r[1] & 0x80) != 0)) {
      throw Exception('R value excessively padded');
    }
    if (lenS > 1 && (input.s[0] == 0x00) && !((input.s[1] & 0x80) != 0)) {
      throw Exception('S value excessively padded');
    }

    Uint8List signature = Uint8List(6 + lenR + lenS);

    // 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S]
    signature[0] = 0x30;
    signature[1] = signature.length - 2;
    signature[2] = 0x02;
    signature[3] = input.r.length;
    signature.setAll(4, input.r);
    signature[4 + lenR] = 0x02;
    signature[5 + lenR] = input.s.length;
    signature.setAll(6 + lenR, input.s);

    return signature;
  }
}


example/dart_bip66_example.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_bip66/dart_bip66.dart';

void main() {
  Uint8List r = Uint8List.fromList(hex.decode(
      '1ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777'));
  Uint8List s = Uint8List.fromList(hex.decode(
      '29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1'));
  // => 304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1
  print(der.encode(DER(r: r, s: s)));
  // => [48, 67, 2, 32, 30, 161, 253, 255, 129, 179, 162, 113, 101, 157, 244, 170, 209, 155, 196, 239, 131, 222, 243, 137, 19, 26, 54, 53, 143, 230, 75, 36, 86, 50, 231, 119, 2, 31, 41, 225, 100, 101, 139, 233, 206, 129, 9, 33, 191, 129, 214, 184, 102, 148, 120, 90, 121, 234, 30, 82, 219, 250, 81, 5, 20, 141, 31, 11, 193]

  Uint8List signature = Uint8List.fromList(hex.decode(
      '304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1'));

  DER decodedSignature = der.decode(signature);
  print(decodedSignature);
  // => 304302201ea1fdff81b3a271659df4aad19bc4ef83def389131a36358fe64b245632e777021f29e164658be9ce810921bf81d6b86694785a79ea1e52dbfa5105148d1f0bc1
  print(der.encode(decodedSignature));
  // => [48, 67, 2, 32, 30, 161, 253, 255, 129, 179, 162, 113, 101, 157, 244, 170, 209, 155, 196, 239, 131, 222, 243, 137, 19, 26, 54, 53, 143, 230, 75, 36, 86, 50, 231, 119, 2, 31, 41, 225, 100, 101, 139, 233, 206, 129, 9, 33, 191, 129, 214, 184, 102, 148, 120, 90, 121, 234, 30, 82, 219, 250, 81, 5, 20, 141, 31, 11, 193]
}
